### DISCUSSION TITLE
Question: Action Notify, problem with Sns message

### DISCUSSION THREAD

**Discussion by SebastianZZ** [cloud-custodian/cloud-custodian#7351] [General]
Hi !

I have created policy with notify action and sns topic in AWS. I want to customize my message because now im getting something looks like a group of random letters. Any idea how to do this ? I can't see any information in docs. 

 policies:
  - name: my-first-policy
    resource: aws.ec2
    filters:
      - "tag:project" : present
      - "State.Name" : running 
    actions:
      - type: notify
        subject: "CUSTODIAN - [{{ account }} {{ region }}]"
        to:
          - cloudcustodian_notify_test
        transport:
          type: sns
          topic: cloudcustodian_notify_test
          region: eu-central-1
          attributes:
            Name: 'AAAAAAAAA'
            Message: 'asdasdasdasdas'
            body: 'asdasdasdasdasda'

**Comment by kapilt** [comment]
tools/c7n_mailer is used for delivering messages, the message sent to the topic or queue from a notify action is a actually a data payload (base64 zlib compressed son). mailer provides for sending to final destinations, looking up email address in ldap, templating messages, etc.

https://cloudcustodian.io/docs/tools/c7n-mailer.html

**Comment by SebastianZZ** [comment]
Yeah I know but I don't want to use SES, i am in Frankfurt and its not available here.

> The to list specifies the intended recipient for the email. You can specify either an email address, an 
SNS topic, a Datadog Metric, or a special value. 

and as u can see, I have added SNS topic
`        to:
          - cloudcustodian_notify_test`

but for some reason it is now working for me. It sens weird msg when i add this topic to transport but i don't know how to specify msg body

**Comment by kapilt** [comment]
mailer is required and can templated messages to a variety of transports ie. sns/smtp/ses/etc

the sns transport for the notify action is a data channel. you have to configure mailer to listen on the other side (sns->sqs->mailer) to get it delivered/rendered for humans.

**Comment by jesayafn** [comment]
The documentation was not clear and needs to be improved. I'm facing the same issue without a solution.

**Comment by ajkerrigan** [comment]
Making the need for c7n-mailer more explicit in #9808 . Suggestions welcome if there's specific wording that would have helped avoid the confusion here.

**Reply by jesayafn** [reply]
https://github.com/cloud-custodian/cloud-custodian/blob/main/tools/c7n_mailer/c7n_mailer/sns_delivery.py

I see the code. Is there support for sending a message through SNS?

**Reply by ajkerrigan** [reply]
Sure, but the message that you send with SNS is still going to be an encrypted/encoded JSON blob rather than a human-readable message. Even with that we contend with breaking up messages to avoid hitting SNS size limits.

If a picture helps, the flow without SNS looks something like this:

```mermaid
flowchart TD
  template[human-readable template] --> c7n-mailer
  notify[policy notify action] --> blob@{shape: lin-doc, label: "encrypted/encoded blob [for transport]"}
  blob --> sqs[SQS queue]
  sqs --> c7n-mailer
  c7n-mailer --> msg@{shape: lin-doc, label: "rendered message [for humans]"}
  msg -- email/slack/etc --> recipient
```

And with SNS it looks more like this:

```mermaid
flowchart TD
  template[human-readable template] --> c7n-mailer
  notify[policy notify action] --> blob@{shape: lin-doc, label: "encrypted/encoded blob [for transport]"}
  blob --> sns[SNS topic]
  sns --> sqs[SQS queue]
  sns --> subs@{shape: processes, label: "other subscribers"}
  sqs --> c7n-mailer
  c7n-mailer --> msg@{shape: lin-doc, label: "rendered message [for humans]"}
  msg -- email/slack/etc --> recipient
```

So while there's a bit of flexibility in the middle, a flow that starts with the `notify` action expects to end with SQS --> c7n-mailer. There's no supported flow for having `notify` deliver to SNS and then direct-subscribing individuals to the topic.
